File: rust/interpreters/src/scanner.rs
use crate::error::Error::{self, CompileTimeError};
use crate::token::{Literal, Token, TokenType, TokenType::*};

pub struct Scanner {
    source: String,
    tokens: Vec<Token>,
    start: usize,
    current: usize,
    line: usize,
}

impl Scanner {
    pub fn new(source: String) -> Scanner {
        Scanner {
            source,
            tokens: vec![],
            start: 0,
            current: 0,
            line: 0,
        }
    }

    pub fn scan_tokens(&mut self) -> Result<Vec<Token>, Error> {
        while !self.is_at_end() {
            self.start = self.current;
            self.scan_token()?
        }
        self.tokens.push(Token::new(EOF, "EOF", None, self.line));
        Ok(self.tokens)
    }

    pub fn scan_token(&mut self) -> Result<(), Error> {
        let c = self.advance();

        match c {
            '(' => self.add_token(LEFT_PAREN, None),
            ')' => self.add_token(RIGHT_PAREN, None),
            '{' => self.add_token(LEFT_BRACE, None),
            '}' => self.add_token(RIGHT_BRACE, None),
            ',' => self.add_token(COMMA, None),
            '.' => self.add_token(DOT, None),
            '-' => self.add_token(MINUS, None),
            '+' => self.add_token(PLUS, None),
            ';' => self.add_token(SEMICOLON, None),
            '*' => self.add_token(STAR, None),
            ':' => self.add_token(COLON, None),
            '?' => self.add_token(QUESTION, None),
            '!' => {
                let tok = if self.matching('=') { BANG_EQUAL } else { BANG };
                self.add_token(tok, None);
            }
            '=' => {
                let tok = if self.matching('=') {
                    EQUAL_EQUAL
                } else {
                    EQUAL
                };
                self.add_token(tok, None);
            }
            '<' => {
                let tok = if self.matching('=') { LESS_EQUAL } else { LESS };
                self.add_token(tok, None);
            }
            '>' => {
                let tok = if self.matching('=') {
                    GREATER_EQUAL
                } else {
                    GREATER
                };
                self.add_token(tok, None);
            }
            '/' => {
                if self.peek_next() == '/' {
                    while self.peek() != '\n' && !self.is_at_end() {
                        self.advance();
                    }
                } else if self.peek_next() == '*' {
                    self.handle_multi_line_comments();
                } else {
                    self.add_token(SLASH, None);
                }
            }
            ' ' | '\r' | '\t' => {}
            '\n' => self.line += 1,
            '"' => self.string()?,
            _ => {
                if self.is_digit(c) {
                    self.number();
                } else if self.is_alpha(c) {
                    self.identifier();
                } else {
                    return Err(CompileTimeError {
                        token: Some(Token::new(
                            TokenType::NIL,
                            &c.clone().to_string().as_str(),
                            None,
                            self.line,
                        )),
                        message: String::from("Unexpected character"),
                    });
                }
            }
        };
        Ok(())
    }

    fn handle_multi_line_comments(&mut self) {
        while !self.is_at_end() {
            if self.peek() == '*' && self.peek_next() == '/' {
                self.advance();
                self.advance();
                break;
            } else {
                self.advance();
            }
        }
    }

    fn is_alpha(&self, c: char) -> bool {
        (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c == '_')
    }

    fn identifier(&mut self) {
        while self.is_alpha_numeric(self.peek()) {
            self.advance();
        }
        let text = self.source[self.start..self.current].to_string();
        let token_type = TokenType::from_string(&text);
        let tt = token_type.unwrap_or(IDENTIFIER);
        self.add_token(tt, None);
    }

    fn is_alpha_numeric(&self, c: char) -> bool {
        self.is_alpha(c) || self.is_digit(c)
    }

    fn number(&mut self) -> Result<(), Error> {
        if !self.is_at_end() {
            while !self.is_at_end() && self.is_digit(self.peek()) {
                self.advance();
            }

            if self.peek() == '.' {
                self.advance();

                while !self.is_at_end() && self.is_digit(self.peek()) {
                    self.advance();
                }
            }

            if self.is_alpha(self.peek_next()) {
                return Err(CompileTimeError {
                    token: Some(Token::new(
                        TokenType::NIL,
                        &self.peek_next().clone().to_string(),
                        None,
                        self.line,
                    )),
                    message: String::from("Unexpected character. check your number"),
                });
            }
        }

        // this is to normalize the advanced current value
        let num = match self.source[self.start..(self.current)]
            .trim()
            .parse::<f64>()
            .ok()
        {
            None => 0.0,
            Some(num) => num,
        };

        self.add_token(NUMBER, Some(Literal::Number(num)));
        Ok(())
    }

    fn peek_next(&self) -> char {
        if self.current + 1 >= self.source.len() {
            return '\0';
        }
        self.get_char_at(self.current)
    }

    fn get_char_at(&self, index: usize) -> char {
        self.source.chars().nth(index).unwrap()
    }

    fn is_digit(&self, c: char) -> bool {
        c >= '0' && c <= '9'
    }

    fn string(&mut self) -> Result<(), Error> {
        while self.peek() != '"' && !self.is_at_end() {
            if self.peek() == '\n' {
                self.line += 1;
            }
            self.advance();
        }
        if self.is_at_end() {
            return Err(CompileTimeError {
                token: Some(Token::new(NIL, "EOF", None, self.line)),
                message: String::from("Unterminated string"),
            });
        }

        self.advance();

        let value = &self.source[self.start + 1..self.current - 1].to_string();
        self.add_token(STRING, Some(Literal::Str(value.to_string())));
        Ok(())
    }

    fn peek(&self) -> char {
        if self.is_at_end() {
            return '\0';
        }
        self.source.chars().nth(self.current).unwrap()
    }

    fn matching(&mut self, expected: char) -> bool {
        if self.is_at_end() {
            return false;
        }

        if self.peek_next() != expected {
            return false;
        }

        self.current += 1;
        true
    }

    fn advance(&mut self) -> char {
        self.current += 1;
        self.source.chars().nth(self.current - 1).unwrap()
    }

    fn add_token(&mut self, token_type: TokenType, literal: Option<Literal>) {
        let text = self.source[self.start..self.current].to_string();
        let text = text.trim();
        let line = self.line;
        self.tokens
            .push(Token::new(token_type, &text, literal, line));
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }
}

File: nim/interpreter/src/scanner.nim
import common_types
import errors
import strutils
import token

# ------- declarations -------
proc isAtEnd(self: Scanner): bool
proc identifier(self: var Scanner)
proc peek(self: Scanner): char
proc peekNext(self: Scanner): char
proc parseStr(self: var Scanner)
proc parseNum(self: var Scanner)
proc match(self: var Scanner, expected: char): bool
proc advance(self: var Scanner): char
proc addToken(self: var Scanner, token_type: TokenType, literal: Literal = Literal(kind: NIL_LIT, nilVal: 0))
# ----------------------------

proc newScanner*(source: string): Scanner=
  return Scanner(source: source, tokens: @[], start: 0, current: 0, line: 0)

proc getKeyword(str: string): TokenType=
  case str:
  of "and": return AND
  of "class": return CLASS
  of "else": return ELSE
  of "false": return FALSE
  of "for": return FOR
  of "fun": return FUN
  of "if": return IF
  of "nil": return NIL
  of "or": return OR
  of "print": return PRINT
  of "return": return RETURN
  of "super": return SUPER
  of "this": return THIS
  of "true": return TRUE
  of "var": return VAR
  of "while": return WHILE
  else: return UNDEFINED

proc matcher(self: var Scanner, c: char, a: TokenType, b: TokenType): TokenType=
    return if self.match(c): a else: b
   
proc scanToken(self: var Scanner)=
  let c = self.advance()

  case c:
  of '(': self.addToken(LEFT_PAREN)
  of ')': self.addToken(RIGHT_PAREN)
  of '{': self.addToken(LEFT_BRACE)
  of '}': self.addToken(RIGHT_BRACE)
  of ',': self.addToken(COMMA)
  of '.': self.addToken(DOT)
  of '-': self.addToken(MINUS)
  of '+': self.addToken(PLUS)
  of ';': self.addToken(SEMICOLON)
  of '*': self.addToken(STAR)
  of '!': self.addToken(self.matcher('=', BANG_EQUAL, BANG))
  of '=': self.addToken(self.matcher('=', EQUAL_EQUAL, EQUAL))
  of '<': self.addToken(self.matcher('=', LESS_EQUAL, LESS))
  of '>': self.addToken(self.matcher('=', GREATER_EQUAL, GREATER))
  of '?': self.addToken(QUESTION)
  of ':': self.addToken(COLON)
  of '/':
    if self.match('/'):
      while (self.peek() != '\n') and (not self.isAtEnd()):
        discard self.advance()
    elif self.match('*'):
      while not self.isAtEnd():
        if self.peek() == '*' and self.peekNext() == '/':
          discard self.advance()
          discard self.advance()
          break
        else: discard self.advance()
    else: self.addToken(SLASH)
  of ' ', '\r', '\t': discard
  of '\n': self.line += 1
  of '"': self.parseStr()
  else:
    if c.isDigit(): self.parseNum()
    elif c.isAlphaAscii() or c == '_': self.identifier()
    else: compileTimeError(newToken(NIL, self.line, $c), "Unexpected character.")

proc scanTokens*(self: var Scanner): seq[Token]=
  while not self.isAtEnd():
    self.start = self.current
    self.scanToken()

  self.tokens.add(newToken(EOF, self.line))
  return self.tokens

proc isAtEnd(self: Scanner): bool=
  return self.current >= self.source.len()

proc identifier(self: var Scanner)=
  while self.peek().isAlphaNumeric():
    discard self.advance()
  let text = self.source.substr(self.start, self.current).strip()
  var token_type = getKeyword(text)
  if token_type == UNDEFINED:
    token_type = IDENTIFIER
  self.addToken(token_type)

proc peek(self: Scanner): char=
  if self.isAtEnd():
    return '\0'
  return self.source[self.current]

proc peekNext(self: Scanner): char=
  if self.current + 1 >= self.source.len():
    return '\0'
  return self.source[self.current + 1]

proc parseStr(self: var Scanner)=
  while self.peek() != '"' and not self.isAtEnd():
    if self.peek() == '\n':
      self.line += 1
    discard self.advance()

  if self.isAtEnd():
    compileTimeError(newToken(EOF, self.line, "EOF"), "Unterminated String")
    return

  discard self.advance()

  let value = self.source.substr(self.start + 1, self.current - 2)
  self.addToken(STRING, Literal(kind: STRING_LIT, strVal: value))

proc parseNum(self: var Scanner)=
  var isFloat = false
  while self.peek().isDigit():
    discard self.advance()
  if self.peek() == '.' and self.peekNext().isDigit():
    isFloat = true
    discard self.advance()
    while self.peek().isDigit():
      discard self.advance()

  let strNum = self.source.substr(self.start, self.current - 1)
  let num = if isFloat: strNum.parseFloat() else: float(strNum.parseInt())
  self.addToken(NUMBER, Literal(kind: NUMBER_LIT, numVal: num))

proc match(self: var Scanner, expected: char): bool=
  if self.isAtEnd() or self.source[self.current] != expected:
    return false

  self.current += 1
  return true

proc advance(self: var Scanner): char=
  self.current += 1
  return self.source[self.current - 1]

proc addToken(self: var Scanner, token_type: TokenType, literal: Literal = Literal(kind: NIL_LIT, nilVal: 0))=
  let text = self.source.substr(self.start, self.current - 1)
  self.tokens.add(newToken(token_type, self.line, text, literal))

